sourcegraph/sourcegraph
# cody: add indexing state for cody app jobs

Close https://github.com/sourcegraph/sourcegraph/issues/54227 https://github.com/sourcegraph/sourcegraph/issues/54230 https://github.com/sourcegraph/sourcegraph/issues/54229

This PR adds:
- New button to add repo into Cody App that will then be indexed automatically by App.
- Indexing progress bar for indexing codebase when connected to App
  - The progress bar will get render on every keystroke entered in the editor
- cleaned up the webmessage for `history` and grouped them together into one function `handleChatHistory`
- Cleaned up onConfigChange functions that caused publishConfig to get called multiple times

### Demos

Add repository to App. it takes App some time to add the repository  and index the repository respectively:
https://www.loom.com/share/80a3b069899e413bafa3396722b1091b
<img width="1229" alt="image" src="https://github.com/sourcegraph/sourcegraph/assets/68532117/59bb3962-151c-4085-bf40-17f951a61138">

Progress bar for indexing job (App only):
https://www.loom.com/share/909a1462dfda41cba583fc58d3c254e5?sid=f2c1fa46-706f-495a-ac10-5b27ec98fcea
<img width="1266" alt="image" src="https://github.com/sourcegraph/sourcegraph/assets/68532117/f5a56f0e-f7cd-437d-b482-9c1ee180a66c">

## Test plan



1. Make sure App is running
2. Connect Cody to App
3. Open a repository in VS Code that is not indexed by Cody App
4. Check the status icon and tooltip to make sure it is showing that the repo(codebase) is not indexed
5. You should also see a button that allows you to `add the repo`
6. Connect Cody to another Instance (ex S2) where the repo is also not indexed, and see if it has the same status icon
7. The `Add the repo` button should not be visible in non-cody-app instance
8. Connect Cody back to the Cody App instance and click on the `Add the repo` button
9. Continue working in your editor, sidebar not included (the status only gets updated everytime you made changes in your current editor, including clicks and keypress ), you should see the status icon and progress bar gets update gradually (it could take some time for the repo to show up in App before it gets sent to the embedding jobs queue, not the editor issue)
10. you should see the progress bar get removed and the status icon changed to indexed when the indexing job is done

Something I didn't test is to make sure this changes would not break the onboarding flow with App*
